File: Student.py
'''
Saahil, James

Student Class
    Functionality:
        Hold information about each student's parking history and information (type of car, strikes, etc.)
        Generate score based on weights
        
        


This class holds information for the student object
'''
import json
import toml
import numpy as np

class Student: 

    #! this code runs at *definition*, so basically when this file is imported.
    with open('distances.json', 'r') as f: DISTANCES = json.load(f)
    with open('config.toml', 'r') as f:   WEIGHTS =   toml.load(f)['weights']
    
    def __init__(self,row):
        '''creates a Student object with instance variables corresponding to the student's google sheet information

        Args:
            row: considering Student objects might be defined in a for loop, the row variable will serve
            to access the corresponding GSheet row and access information from it

            day: the day of the week the program is being run on
                   
        
        Actions:
            uses GSheet class to define instance variables (name, type of car, weights, etc.)

        '''

        self.data = {
            'name':    row[0],
            'big_car': row[1],
            'zipcode': row[2],
            'crit':    row[3:8],
            'sports':  row[8:13],
            'carpool': row[13:15], #seniors=13, youngns=14
            'free':    row[15:17], #last period: 15, first: 16
            'par':     row[17],
            'strikes': row[18]
        }
        #the amount of days it has been since the student has been able to park on campus - adjust score based on this variable in generateScore()

    def __repr__(self):
        return 'Student: ' + self.name

    def getName(self):
        return self.name

    def canParallelPark(self):
        return bool(self.parallel)

    def hasSmallCar(self):
        return not self.car
        

    def __hash__(self):
        """hash implementation for Student. Don't call, use `hash(Student)`. Based on the name.
        Returns:
            int: the hash for this Student
        """
        return hash(self.name)
    
    def generateScore(self,day):
        '''uses weights to return a score that the Sorter function/class will use to assign the Student a parking zone'''

        student_attrs = np.array([self.data['free'] + 
                                [self.data['sports'][day]] + 
                                [self.data['crit'][day]] + 
                                [self.data['strikes']]])

        weights = np.array([[Student.WEIGHTS[i] for i in [
            "First_Period_Free",
            "Last_Period_Free",
            "Sports_Away_Game",
            "Critical_Need",
            "Strike",
            "Crash"]]])

        score = student_attrs.dot(weights.T) #take dot product (need to transpose because dot is just matmul so u dot v needs to be [u][v]^T)

        #* multiply the score by a the carpool multiplier
        score *= (
                        1 +                                                                                 # start with one, so by default it's x1, so we get same score
                        self.data['carpool'][day]*Student.WEIGHTS['Carpool_Multiplier_Per_Non_Senior'] +    # total minus seniors is num underclassmen, mult by 0.25 bc we weight underclassmen less
                        self.data['carpool'][day]*Student.WEIGHTS['Carpool_Multiplier_Per_Senior']          # add one multiplication per senior, because you're freeing up another spot
                    )
        
        score += np.random.normal(0, Student.WEIGHTS['Random_Factor_STDEV']) #add random stuff, mean 0
        
        return score + Student.distScore(str(self.data['zipcode']))

    @staticmethod
    def distScore(zipcode):
        """gets the score for a given zip code.

        Args:
            zipcode (int): the zip code to get the score for

        Raises:
            KeyError: if the zipcode is invalid
            
        Returns:
            float: the resulting score, based on the l1 distance to 94010 (school)
        """
        try:
            return Student.DISTANCES[zipcode]*0.9
        except: #hehe general except go brrrrrrrrr (i don't want to put KeyError cause idk if that's what this raises, idek if this is a dict or json object or what)
            return 0 #not a valid zipcode, their fault, score 0

File: parking/src/Student.rs
use std::collections::HashMap;
use std::fs;
use std::error::Error;
use serde_json::{Value, Map};
use rand::distributions::{Normal, Distribution};

pub struct Student{
    name: str,
    score: f32,
    row: Vec<u8>,
}

impl Student{
    const DISTANCES: Map<String, Value> = Student::read_config("~/parking/distance.json").unwrap();
    
    fn read_config(path: &str) -> Result<Map<String, Value>, Box<Error>> {
        let config = fs::read_to_string(path)?;
        let parsed: f32 = serde_json::from_str(&config)? as f32;
        let obj: Map<String, f32> = parsed.as_object().unwrap().clone();
        Ok(obj)
    }

    pub fn new(student_name: &str, row: &Vec<i32>) -> Student{
        Student {
            name: *student_name,
            row: row.clone(),
        }
    }

    pub fn generate_score(&self, day: usize) -> f32{
        let crit: u8 = self.row[2..7][day];
        let sports: u8 = self.row[7..12][day];
        let fp_free = self.row[14];
        let lp_free = self.row[15];
        let dist: f32 = Student::DISTANCES[self.row[1]];
        let carpool_seniors: u8 = self.row[12];
        let carpool_youngns: u8 = self.row[13];
        let strikes = self.row[17]
        let weights: Vec<f32> = vec![16.0, 8.0, 10.0, 40.0, -20];
        let mut score: f32 = weights.iter().zip(vec![fp_free, lp_free, crit, sports, strikes].iter()).map(|(x, y)| x * (*y as f32)).sum();
        score *= 1.0 + carpool_seniors as f32 + (carpool_youngns as f32 * 0.25);
        score += Normal::new(0.0, 5.0);

        score
    }

    pub fn can_parallel_park(&self) -> bool{
        self.row[16] == 1
    }
    pub fn has_small_car(&self) -> bool{
        self.row[0] == 1
    }
    pub fn get_name(&self) -> String{
        String::from(&self.name)
    }

}